### PR TITLE
DAOS-2344 vos: Add flags to valid umem_off_t as well

### PR DESCRIPTION
**Pull request by jolivier23**
For my punch log change, I need a bit in a valid umem_off_t
to indicate a punch to avoid wasting additional space.  This
patch modifies the old invalid API to allow setting flags on
any umem_off_t.   It modifies the change to use the top bits
of the address (should also be safe) so we don't conflict
with any offset bits in valid offsets.
Additionally, this patch removes all public API references to
umem_id_t and exports the generic umem API for umem_off_t only
since this is how it is used.   This simplifies the API and
may reduce future confusion or misuse.

### CODE REVIEW

File: src/common/mem.c [hunk 1]
```diff
@@ -37,52 +37,72 @@
...
 {
-	return OID_EQUALS(ummid1, ummid2);
+	if (OID_IS_NULL(oid))
+		return UMOFF_NULL;
+
+	return oid.off;
 }
 
+struct umem_tx_stage_item {
+	int		 txi_magic;
```
Comment (ryon-jensen) [inline src/common/mem.c]: Probably obvious to most, but what is "magic"?

File: src/common/mem.c [hunk 2]
```diff
@@ -338,44 +356,20 @@ umem_tx_errno(int err)
...
+umem_off_t
 vmem_alloc(struct umem_instance *umm, size_t size, uint64_t flags,
 	   unsigned int type_num)
 {
-	umem_id_t ummid = UMMID_NULL;
-
-	ummid.off = (uint64_t)(flags & POBJ_FLAG_ZERO ?
-			       calloc(1, size) : malloc(size));
-	return ummid;
+	return (uint64_t)((flags & POBJ_FLAG_ZERO) ?
```
Comment (ryon-jensen) [inline src/common/mem.c]: nit - I know umem_off_t is a uint64_t, but for consistency, can you make the return type and the cast type the same? :)

Comment (jolivier23) [reply src/common/mem.c]: yeah, fair point.   I think it's just cause I copied it from the old code :-)

File: src/common/mem.c (hunk 1 above)
Comment (jolivier23) [reply src/common/mem.c]: @ryon-jensen It's used for sanity check that the item is a valid one.  Kind of like a weak type check.  It's set when adding a callback and checked when using it with D_ASSERT

**Pull request merged**